src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -243,30 +243,23 @@ BoundExpression MakeIsDeclarationPattern(SyntaxNode syntax, BoundExpression lowe
             }
             else // type parameter or value type
             {
-                // bool Is<T>(this object i, out T o)
+                // bool Is<T>(this object i, ref T o)
                 // {
                 //     // inefficient because it performs the type test twice, and also because it boxes the input.
-                //     bool s;
-                //     o = (s = i is T) ? (T)i : default(T);
-                //     return s;
+                //     return i is T && (o = (T)i; true);
                 // }
 
                 // Because a cast involving a type parameter is not necessarily a valid conversion (or, if it is, it might not
                 // be of a kind appropriate for pattern-matching), we use `object` as an intermediate type for the input expression.
                 var objectType = _factory.SpecialType(SpecialType.System_Object);
-                var s = _factory.SynthesizedLocal(_factory.SpecialType(SpecialType.System_Boolean), syntax);
                 var i = _factory.SynthesizedLocal(objectType, syntax); // we copy the input to avoid double evaluation
-                return _factory.Sequence(
-                    ImmutableArray.Create(s, i),
-                    ImmutableArray.Create<BoundExpression>(
-                        _factory.AssignmentExpression(_factory.Local(i), _factory.Convert(objectType, loweredInput)),
-                        _factory.AssignmentExpression(loweredTarget, _factory.Conditional(
-                            _factory.AssignmentExpression(_factory.Local(s), _factory.Is(_factory.Local(i), type)),
-                            _factory.Convert(type, _factory.Local(i)),
-                            _factory.Default(type), type))
-                        ),
-                    _factory.Local(s)
-                    );
+                return _factory.MakeSequence(i,
+                    _factory.AssignmentExpression(_factory.Local(i), _factory.Convert(objectType, loweredInput)),
```

Comment:
Why box before the test? You could perform the test on the original value and only box to perform the conversion when the test succeeds.

Reply:
I tried it. for some reason it has *bigger* codegen (with one additional box).

Test case from https://github.com/dotnet/roslyn/issues/19153:

- current PR:
```cs
{
	// Method begins at RVA 0x21ac
	// Code size 44 (0x2c)
	.maxstack 1
	.locals init (
		[0] !!U,
		[1] !!T,
		[2] object
	)

	IL_0000: ldarg.0
	IL_0001: stloc.0
	IL_0002: ldloc.0
	IL_0003: box !!U
	IL_0008: brfalse.s IL_002b

	IL_000a: ldloc.0
	IL_000b: box !!U
	IL_0010: stloc.2
	IL_0011: ldloc.2
	IL_0012: isinst !!T
	IL_0017: brfalse.s IL_002b

	IL_0019: ldloc.2
	IL_001a: unbox.any !!T
	IL_001f: stloc.1
	IL_0020: ldloc.1
	IL_0021: box !!T
	IL_0026: call void [mscorlib]System.Console::Write(object)

	IL_002b: ret
} // end of method P::M
```
- move box to the second assignment:
```cs
{
	// Method begins at RVA 0x21ac
	// Code size 49 (0x31)
	.maxstack 1
	.locals init (
		[0] !!U,
		[1] !!T,
		[2] !!U //! loweredInput.Type
	)

	IL_0000: ldarg.0
	IL_0001: stloc.0
	IL_0002: ldloc.0
	IL_0003: box !!U
	IL_0008: brfalse.s IL_0030

	IL_000a: ldloc.0
	IL_000b: stloc.2
	IL_000c: ldloc.2
	IL_000d: box !!U
	IL_0012: isinst !!T
	IL_0017: brfalse.s IL_0030

	IL_0019: ldloc.2
	IL_001a: box !!U
	IL_001f: unbox.any !!T
	IL_0024: stloc.1
	IL_0025: ldloc.1
	IL_0026: box !!T
	IL_002b: call void [mscorlib]System.Console::Write(object)

	IL_0030: ret
} // end of method P::M
```

Reply:
I think `isinst` needs the box anyways.

Reply:
I don’t think isinst needs a box. I’d rather have larger code gen that does not box.

Reply:
This is the relevant code,
```cs
	IL_000b: box !!U
	IL_0010: stloc.2
	IL_0011: ldloc.2
	IL_0012: isinst !!T
```
vs.
```cs
	IL_000c: ldloc.2
	IL_000d: box !!U
	IL_0012: isinst !!T
```
In the first example we capture the box so isinst doesn't need to box again, Notice one additional box in the second example.

Reply:
isinst  does require boxing.  

Formally only objects know their type and thus their instances can be type-checked. 

Smart JIT may elide/fuse “box;isinst;” sequences since at jit time the type of T is known exactly or known to be a reference type.

Reply:
I think it should be checked whether storing the boxed value into a temp does not result in defeating jit optimizations and performing actual boxing.